modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java

```diff
@@ -166,18 +159,29 @@ private void doFlush(ChannelHandlerContext ctx) {
                     }
                 });
             }
-            ctx.flush();
-            if (channel.isActive() == false) {
-                failQueuedWrites();
-                return;
+            if (channel.isWritable()) {
```

Comment:
I'm suspicious about checking this flag twice (once here and once in the loop condition), can we reduce it to one read?

Indeed given the possibility to become unwritable at any time (on a different thread) do we need to check this at all? I mean we could become unwritable after this check but before calling `ctx.flush()` on the next line.

Reply:
> I'm suspicious about checking this flag twice (once here and once in the loop condition), can we reduce it to one read?

We do need to read twice in case we try to flush inside the loop to figure out if the flush made us writable again or not.
We could technically not check twice in case the channel was writable when checking inside the loop, but I couldn't really find a non-confusing way to do so and figured since it's just a cheap volatile read it doesn't really matter. Since we're now not missing a flush any more a concurrent change to not-writable between two checks just means that the next flush task enqueued after the currently executing task will make us writable again, then the writability changed hook will trigger another round of `doFlush` and deal with the write and we're all good I think :)